rust-client-sdk/src/lib.rs

```diff
@@ -69,17 +72,67 @@ impl HathoraClient {
         &self,
         token: &str,
         state_id: &str,
-    ) -> Result<WebSocket<MaybeTlsStream<TcpStream>>> {
-        let websocket_url = format!("wss://{}/connect/{}", self.coordinator_host, self.app_id);
-        let (mut socket, _response) =
-            connect(Url::parse(&websocket_url).unwrap()).expect("Can't connect to websockets");
+        transport_type: HathoraTransportType,
+    ) -> Result<Box<dyn HathoraTransport>> {
+        match transport_type {
+            HathoraTransportType::WebSocket => {
+                let mut transport = WebsocketTransport::new(
+                    self.app_id.clone(),
+                    Some(self.coordinator_host.clone()),
+                );
+                transport.connect(state_id, token)?;
+                Ok(Box::new(transport))
+            }
+        }
+    }
+
+    pub fn get_user_from_token(token: &str) -> Result<String> {
+        let segments: Vec<&str> = token.split('.').collect();
+        let bytes = base64::decode_config(segments[1], base64::URL_SAFE_NO_PAD)?;
+        let token: Token = serde_json::from_slice(&bytes)?;
+        Ok(token.id)
+    }
+}
+
+pub enum HathoraTransportType {
+    WebSocket,
+}
+
+pub trait HathoraTransport {
+    fn connect(&mut self, state_id: &str, token: &str) -> Result<()>;
+
+    fn write(&mut self, data: Vec<u8>) -> Result<()>;
+
+    fn read(&mut self) -> Result<Vec<u8>>;
```

Comment:
I think I prefer `read_message()` and `write_message()`, since we handle all the underlying message parsing stuff for you